File: 316-remove-duplicate-letters/remove-duplicate-letters.java
import java.util.*;

class Solution {
    public String removeDuplicateLetters(String s) {

        int[] last = new int[26];

        // Store last position
        for (int i = 0; i < s.length(); i++) {
            last[s.charAt(i) - 'a'] = i;
        }

        Stack<Character> stack = new Stack<>();
        boolean[] used = new boolean[26];

        for (int i = 0; i < s.length(); i++) {

            char ch = s.charAt(i);
            int index = ch - 'a';

            // Already present
            if (used[index]) {
                continue;
            }

            // Remove bigger characters
            // only if they appear again later
            while (
                !stack.isEmpty() &&
                stack.peek() > ch &&
                last[stack.peek() - 'a'] > i
            ) {
                char removed = stack.pop();
                used[removed - 'a'] = false;
            }

            stack.push(ch);
            used[index] = true;
        }

        StringBuilder ans = new StringBuilder();

        for (char ch : stack) {
            ans.append(ch);
        }

        return ans.toString();
    }
}

File: 316-remove-duplicate-letters/remove-duplicate-letters.py
class Solution:
    def removeDuplicateLetters(self, s: str) -> str:

        last = {}

        # Store last position of each character
        for i in range(len(s)):
            last[s[i]] = i

        stack = []
        used = set()

        for i in range(len(s)):
            ch = s[i]

            # Already present
            if ch in used:
                continue

            # Remove bigger characters
            # only if they appear again later
            while (
                stack
                and stack[-1] > ch
                and last[stack[-1]] > i
            ):
                removed = stack.pop()
                used.remove(removed)

            stack.append(ch)
            used.add(ch)

        return ''.join(stack)